sentry/src/main/java/io/sentry/MetricsAggregator.java

```diff
@@ -0,0 +1,320 @@
+    for (final @NotNull Map.Entry<String, String> defaultTag : defaultTags.entrySet()) {
+      final @NotNull String key = defaultTag.getKey();
+      if (!enrichedTags.containsKey(key)) {
+        enrichedTags.put(key, defaultTag.getValue());
+      }
+    }
+    return enrichedTags;
+  }
+
+  @Override
+  public void flush(final boolean force) {
+    final @NotNull Set<Long> flushableBuckets = getFlushableBuckets(force);
+    if (flushableBuckets.isEmpty()) {
+      logger.log(SentryLevel.DEBUG, "Metrics: nothing to flush");
+      return;
+    }
+    logger.log(SentryLevel.DEBUG, "Metrics: flushing " + flushableBuckets.size() + " buckets");
+
+    final @NotNull StringBuilder writer = new StringBuilder();
+    for (long bucketKey : flushableBuckets) {
+      final @Nullable Map<String, Metric> metrics = buckets.remove(bucketKey);
+      if (metrics != null) {
+        MetricsHelper.encodeMetrics(bucketKey, metrics.values(), writer);
+      }
+    }
+
+    if (writer.length() == 0) {
+      logger.log(SentryLevel.DEBUG, "Metrics: only empty buckets found");
+      return;
+    }
+
+    logger.log(SentryLevel.DEBUG, "Metrics: capturing metrics");
+    final @NotNull EncodedMetrics encodedMetrics = new EncodedMetrics(writer.toString());
+    hub.captureMetrics(encodedMetrics);
+  }
+
+  @NotNull
+  private Set<Long> getFlushableBuckets(final boolean force) {
+    if (force) {
+      return buckets.keySet();
+    } else {
+      // get all keys, including the cutoff key
+      final long cutoffTimestampMs =
+          MetricsHelper.getCutoffTimestampMs(timeProvider.getTimeMillis());
+      final long cutoffKey = MetricsHelper.getTimeBucketKey(cutoffTimestampMs);
+      return buckets.headMap(cutoffKey, true).keySet();
+    }
+  }
+
+  @SuppressWarnings("Java8MapApi")
+  @NotNull
+  private Map<String, Metric> getOrAddTimeBucket(final long bucketKey) {
+    @Nullable Map<String, Metric> bucket = buckets.get(bucketKey);
+    if (bucket == null) {
+      // although buckets is thread safe, we still need to synchronize here to avoid creating
+      // the same bucket at the same time
+      synchronized (buckets) {
+        bucket = buckets.get(bucketKey);
+        if (bucket == null) {
+          bucket = new HashMap<>();
+          buckets.put(bucketKey, bucket);
+        }
+      }
+    }
+    return bucket;
+  }
+
+  @Override
+  public void close() throws IOException {
+    synchronized (this) {
+      isClosed = true;
+      executorService.close(0);
+    }
+    flush(true);
+  }
+
+  @Override
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public void run() {
+    flush(false);
+
+    synchronized (this) {
+      if (!isClosed) {
+        executorService.schedule(this, MetricsHelper.FLUSHER_SLEEP_TIME_MS);
+      }
+    }
+  }
+
+  @TestOnly
+  void setTimeProvider(final @NotNull TimeProvider timeProvider) {
+    this.timeProvider = timeProvider;
+  }
+
+  public interface TimeProvider {
```

Comment:
can we use a `SentryDateProvider`, using `DateUtils.nanosToMillis(sentryDate.nanoTimestamp())` or by adding a `millisTimestamp` method to `SentryDate`?

Reply:
`DateUtils.nanosToMillis` returns a double, so I can't use that. I'm resorting to `TimeUnit.NANOSECONDS.toMillis(dateProvider.now().nanoTimestamp());`